File: dailyquest_web/src/test/java/dailyquest/quest/repository/QuestLogRepositoryUnitTest.java
package dailyquest.quest.repository;

import dailyquest.config.JpaAuditingConfiguration;
import dailyquest.quest.dto.QuestLogSearchCondition;
import dailyquest.quest.dto.QuestStatisticsResponse;
import dailyquest.quest.entity.Quest;
import dailyquest.quest.entity.QuestLog;
import dailyquest.quest.entity.QuestState;
import dailyquest.quest.entity.QuestType;
import dailyquest.user.entity.ProviderType;
import dailyquest.user.entity.User;
import dailyquest.user.repository.UserRepository;
import org.junit.jupiter.api.DisplayName;
import org.junit.jupiter.api.Test;
import org.mockito.Answers;
import org.mockito.MockedStatic;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.boot.test.autoconfigure.orm.jpa.DataJpaTest;
import org.springframework.context.annotation.Import;

import java.time.LocalDate;
import java.time.LocalTime;
import java.util.List;

import static org.assertj.core.api.Assertions.assertThat;
import static org.mockito.Mockito.mockStatic;
import static org.mockito.Mockito.when;

@DisplayName("퀘스트 로그 리포지토리 유닛 테스트")
@DataJpaTest
@Import(JpaAuditingConfiguration.class)
public class QuestLogRepositoryUnitTest {

    @Autowired
    QuestLogRepository questLogRepository;

    @Autowired
    UserRepository userRepository;

    @Autowired
    QuestRepository questRepository;

    @DisplayName("퀘스트 통계 조회 시 타입과 상태 별로 그룹화된 카운트가 반환된다")
    @Test
    public void testGetQuestLog() throws Exception {
        //given
        User user = new User("", "user", ProviderType.GOOGLE);
        User savedUser = userRepository.save(user);

        int completeCount = 3;
        int failCount = 2;
        int discardCount = 5;

        int mainCount = completeCount + failCount;

        LocalDate today = LocalDate.now();
        LocalTime mockTime = LocalTime.of(8, 0);

        try(MockedStatic<LocalTime> ignored = mockStatic(LocalTime.class, Answers.CALLS_REAL_METHODS)) {
            when(LocalTime.now()).thenReturn(mockTime);

            for (int i = 0; i < completeCount; i++) {
                Quest quest = new Quest("quest", "desc", savedUser, 1, QuestState.COMPLETE, QuestType.MAIN, null, null);
                Quest savedQuest = questRepository.save(quest);
                questLogRepository.save(new QuestLog(savedQuest));
            }

            for (int i = 0; i < failCount; i++) {
                Quest quest = new Quest("quest", "desc", savedUser, 1, QuestState.FAIL, QuestType.MAIN, null, null);
                Quest savedQuest = questRepository.save(quest);
                questLogRepository.save(new QuestLog(savedQuest));
            }

            for (int i = 0; i < discardCount; i++) {
                Quest quest = new Quest("quest", "desc", savedUser, 1, QuestState.DISCARD, QuestType.SUB, null, null);
                Quest savedQuest = questRepository.save(quest);
                questLogRepository.save(new QuestLog(savedQuest));
            }

            Quest quest = new Quest("quest", "desc", savedUser, 1, QuestState.PROCEED, QuestType.SUB, null, null);
            Quest savedQuest = questRepository.save(quest);
            questLogRepository.save(new QuestLog(savedQuest));

            //when
            List<QuestStatisticsResponse> groupedQuestLogs = questLogRepository.getGroupedQuestLogs(savedUser.getId(), new QuestLogSearchCondition());

            //then
            assertThat(groupedQuestLogs).allMatch(log -> log.getLoggedDate().equals(today));
            assertThat(groupedQuestLogs).anyMatch(log -> log.getCompleteCount() == completeCount);
            assertThat(groupedQuestLogs).anyMatch(log -> log.getFailCount() == failCount);
            assertThat(groupedQuestLogs).anyMatch(log -> log.getDiscardCount() == discardCount);
            assertThat(groupedQuestLogs).anyMatch(log -> log.getMainCount() == mainCount);
            assertThat(groupedQuestLogs).anyMatch(log -> log.getMainCount() == discardCount);
        }
    }

    @DisplayName("퀘스트 통계 조회 시 조회되는 값이 없으면 빈 리스트가 반환된다")
    @Test
    public void testEmptyQuestLog() throws Exception {
        //given
        User user = new User("", "user", ProviderType.GOOGLE);
        User savedUser = userRepository.save(user);

        //when
        List<QuestStatisticsResponse> groupedQuestLogs = questLogRepository.getGroupedQuestLogs(savedUser.getId(), new QuestLogSearchCondition());

        //then
        assertThat(groupedQuestLogs).isEmpty();
    }

}

File: batch/src/test/unit/dailyquest/quest/repository/QuestLogRepositoryUnitTest.kt
package dailyquest.quest.repository

import dailyquest.config.QueryDslConfig
import dailyquest.quest.entity.QuestState
import dailyquest.quest.entity.QuestState.*
import io.mockk.junit5.MockKExtension
import jakarta.persistence.EntityManager
import org.assertj.core.api.Assertions.assertThat
import org.junit.jupiter.api.DisplayName
import org.junit.jupiter.api.Nested
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.extension.ExtendWith
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.boot.test.autoconfigure.orm.jpa.DataJpaTest
import org.springframework.context.annotation.Import
import org.springframework.data.domain.Pageable
import java.time.LocalDate

@Import(QueryDslConfig::class)
@ExtendWith(MockKExtension::class)
@DisplayName("퀘스트 로그 리포지토리 유닛 테스트")
@DataJpaTest
class QuestLogRepositoryUnitTest @Autowired constructor(
    private val questLogRepository: QuestLogRepository,
    private val entityManager: EntityManager
) {

    @DisplayName("getAllUserIdsWhoAchievedPerfectDay 메서드 호출 시")
    @Nested
    inner class TestGetAllUserIdsWhoAchievedPerfectDay {
        @DisplayName("loggedDate에 등록한 퀘스트를 모두 완료한 유저 아이디만 조회된다")
        @Test
        fun `loggedDate에 등록한 퀘스트를 모두 완료한 유저 아이디만 조회된다`() {
            //given
            val loggedDate = LocalDate.of(2020, 1, 1)
            val shouldBeContainedUserId = 1L
            saveQuestLog(shouldBeContainedUserId, 1L, PROCEED, loggedDate)
            saveQuestLog(shouldBeContainedUserId, 1L, COMPLETE, loggedDate)

            saveQuestLog(2L, 2L, PROCEED, loggedDate)

            //when
            val result = questLogRepository.getAllUserIdsWhoAchievedPerfectDay(loggedDate, Pageable.unpaged())

            //then
            assertThat(result).containsExactly(shouldBeContainedUserId)
        }

        @DisplayName("동일한 퀘스트에 대해 PROCEED와 COMPLETE 날짜가 다른 경우 조회되지 않는다")
        @Test
        fun `동일한 퀘스트에 대해 PROCEED와 COMPLETE 날짜가 다른 경우 조회되지 않는다`() {
            //given
            val loggedDate = LocalDate.of(2020, 1, 1)
            val shouldNotBeContainedUserId = 1L
            saveQuestLog(shouldNotBeContainedUserId, 1L, PROCEED, loggedDate)
            saveQuestLog(shouldNotBeContainedUserId, 1L, COMPLETE, loggedDate.plusDays(1))
            saveQuestLog(shouldNotBeContainedUserId, 2L, PROCEED, loggedDate.minusDays(1))
            saveQuestLog(shouldNotBeContainedUserId, 2L, COMPLETE, loggedDate)

            //when
            val result = questLogRepository.getAllUserIdsWhoAchievedPerfectDay(loggedDate, Pageable.unpaged())

            //then
            assertThat(result).doesNotContain(shouldNotBeContainedUserId)
        }

        @DisplayName("조회일에 포기하지 않은 모든 퀘스트를 완료한 경우에도 조회되지 않는다")
        @Test
        fun `조회일에 포기하지 않은 모든 퀘스트를 완료한 경우에도 조회되지 않는다`() {
            //given
            val loggedDate = LocalDate.of(2020, 1, 1)
            val shouldNotBeContainedUserId = 1L
            saveQuestLog(shouldNotBeContainedUserId, 1L, PROCEED, loggedDate)
            saveQuestLog(shouldNotBeContainedUserId, 1L, DISCARD, loggedDate)
            saveQuestLog(shouldNotBeContainedUserId, 2L, PROCEED, loggedDate)
            saveQuestLog(shouldNotBeContainedUserId, 2L, COMPLETE, loggedDate)

            //when
            val result = questLogRepository.getAllUserIdsWhoAchievedPerfectDay(loggedDate, Pageable.unpaged())

            //then
            assertThat(result).doesNotContain(shouldNotBeContainedUserId)
        }

        @DisplayName("조회일에 삭제하지 않은 모든 퀘스트를 완료한 경우에도 조회되지 않는다")
        @Test
        fun `조회일에 삭제하지 않은 모든 퀘스트를 완료한 경우에도 조회되지 않는다`() {
            //given
            val loggedDate = LocalDate.of(2020, 1, 1)
            val shouldNotBeContainedUserId = 1L
            saveQuestLog(shouldNotBeContainedUserId, 1L, PROCEED, loggedDate)
            saveQuestLog(shouldNotBeContainedUserId, 1L, DELETE, loggedDate)
            saveQuestLog(shouldNotBeContainedUserId, 2L, PROCEED, loggedDate)
            saveQuestLog(shouldNotBeContainedUserId, 2L, COMPLETE, loggedDate)

            //when
            val result = questLogRepository.getAllUserIdsWhoAchievedPerfectDay(loggedDate, Pageable.unpaged())

            //then
            assertThat(result).doesNotContain(shouldNotBeContainedUserId)
        }

        private fun saveQuestLog(userId: Long, questId: Long, state: QuestState, loggedDate: LocalDate) {
            val query =
                entityManager.createNativeQuery("insert into quest_log (quest_id, user_id, state, logged_date, type, created_date, last_modified_date) values (?, ?, ?, ?, 'MAIN', now(), now())")
            query.setParameter(1, userId)
            query.setParameter(2, questId)
            query.setParameter(3, state.name)
            query.setParameter(4, loggedDate)
            query.executeUpdate()
        }
    }
}